src/expr/src/relation/join_input_mapper.rs

```diff
@@ -0,0 +1,356 @@
+    arities: Vec<usize>,
+    /// Looks up which input each column belongs to. Derived from `arities`.
+    /// Stored as a field to avoid recomputation.
+    input_relation: Vec<usize>,
+    /// The sum of the arities of the previous inputs in the join. Derived from
+    /// `arities`. Stored as a field to avoid recomputation.
+    prior_arities: Vec<usize>,
+}
+
+impl JoinInputMapper {
+    /// Creates a new `JoinInputMapper` and calculates the mapping of global context
+    /// columns to local context columns.
+    pub fn new(inputs: &[RelationExpr]) -> Self {
+        let types = inputs.iter().map(|i| i.typ()).collect::<Vec<_>>();
+        let arities = types
+            .iter()
+            .map(|t| t.column_types.len())
+            .collect::<Vec<_>>();
+
+        let mut offset = 0;
+        let mut prior_arities = Vec::new();
+        for input in 0..inputs.len() {
+            prior_arities.push(offset);
+            offset += arities[input];
+        }
+
+        let input_relation = arities
+            .iter()
+            .enumerate()
+            .flat_map(|(r, a)| std::iter::repeat(r).take(*a))
+            .collect::<Vec<_>>();
+
+        JoinInputMapper {
+            arities,
+            input_relation,
+            prior_arities,
+        }
+    }
+
+    /// All column numbers in order for a particular input in the local context
+    #[inline]
+    pub fn local_columns(&self, index: usize) -> Range<usize> {
+        0..self.arities[index]
+    }
+
+    /// All column numbers in order for a particular input in the global context
+    #[inline]
+    pub fn global_columns(&self, index: usize) -> Range<usize> {
+        self.prior_arities[index]..(self.prior_arities[index] + self.arities[index])
+    }
+
+    /// Takes an expression from the global context and creates a new version
+    /// where column references have been remapped to the local context.
+    /// Assumes that all columns in `expr` are from the same input.
+    pub fn map_expr_to_local(&self, expr: &ScalarExpr) -> ScalarExpr {
+        let mut expr = expr.clone();
+        expr.visit_mut(&mut |e| {
+            if let ScalarExpr::Column(c) = e {
+                *c -= self.prior_arities[self.input_relation[*c]];
+            }
+        });
+        expr
+    }
+
+    /// Takes an expression from the local context of the `index`th input and
+    /// creates a new version where column references have been remapped to the
+    /// global context.
+    pub fn map_expr_to_global(&self, expr: &ScalarExpr, index: usize) -> ScalarExpr {
+        let mut expr = expr.clone();
+        expr.visit_mut(&mut |e| {
+            if let ScalarExpr::Column(c) = e {
+                *c += self.prior_arities[index];
+            }
+        });
+        expr
+    }
+
+    /// Remap column numbers from the global to the local context.
+    /// Assumes column numbers are from the same input.
+    pub fn map_columns_to_local(&self, columns: &[usize]) -> Vec<usize> {
```

Comment:
Does this need to return a `Vec<_>` or could it return an `impl Iter<Item=usize>`? More "idiomatic Rust" comments, and not real issues that should block anything!

Reply:
Yes, it could return an `impl Iter<Item=usize>`, but I just haven't figured out how to properly mark the lifetimes and avoid the
```
error[E0373]: closure may outlive the current function, but it borrows `self`, which is owned by the current function
```

Reply:
I see! I think `impl Iter<Item=usize>+'a` should work, with a `'a` added in
```rust
pub fn map_columns_to_local<'a>(&'a self, columns: &'a [usize])
```
but I could be wrong!

Reply:
Ah it turns out the uses of `into_iter()` was for the method `lookup_inputs`, for which we could return an iterator without marking the method with lifetimes. Kept `map_columns_to_local` as returning `Vec<usize>` for now and changed the return value of `lookup_inputs`.